File: rust/43.multiply-strings.rs
/*
 * @lc app=leetcode id=43 lang=rust
 *
 * [43] Multiply Strings
 */

// @lc code=start
impl Solution {
    pub fn multiply(num1: String, num2: String) -> String {
        let zero = String::from("0");
        if num1 == zero || num2 == zero {
            return zero;
        }
        let mut result = vec![0; num1.len() + num2.len()];
        for (i, n1) in num1.bytes().rev().enumerate() {
            let n1 = (n1 - b'0') as i32;
            for (j, n2) in num2.bytes().rev().enumerate() {
                let n2 = (n2 - b'0') as i32;
                let r = n1 * n2 + result[i + j];
                result[i + j] = r % 10;
                result[i + j + 1] += r / 10;
            }
        }
        result
            .into_iter()
            .rev()
            .map(|item| item.to_string())
            .collect::<String>()
            .trim_start_matches('0')
            .to_string()
    }
}
// @lc code=end

File: javascript/43.multiply-strings.js
/*
 * @lc app=leetcode id=43 lang=javascript
 *
 * [43] Multiply Strings
 */

// @lc code=start
/**
 * @param {string} num1
 * @param {string} num2
 * @return {string}
 */
var multiply = function (num1, num2) {
  const zero = "0";
  if (num1 === zero || num2 === zero) return zero;
  const d = zero.charCodeAt();
  const num1len = num1.length;
  const num2len = num2.length;
  const result = new Array(num1len + num2len).fill(0);
  for (let i = num1len - 1; i >= 0; i--) {
    const n1 = num1[i].charCodeAt() - d;
    for (let j = num2len - 1; j >= 0; j--) {
      const n2 = num2[j].charCodeAt() - d;
      const r = n1 * n2 + result[i + j + 1];
      result[i + j + 1] = r % 10;
      result[i + j] += ~~(r / 10);
    }
  }
  return result.join("").replace(/^0+/, "");
};
// @lc code=end
